src/doc/languagespec.tex

```diff
@@ -5913,6 +5913,19 @@ \section{Dictionary Lookups}
                                 character, it means to exclude that set).
 \end{tabular}
 
+Optional parameters are passed by passing the parameter string as an 
+argument, then the corresponding value as the subsequent argument.  
+For example, setting the \qkw{mindist} as 0.5 and \qkw{maxdist} 
+as 1.0 is done like so:
+
+\begin{code}
+    trace (pos, dir, "mindist", 0.5, "maxdist", 1.0);
+\end{code}
+
+Furthermore, \qkw{mindist} and \qkw{maxdist} should be viewed as scalars,
+which multiplied by the magnitude of the direction vector is the length of the
```

Comment:
Is that true? If `dir` is not unit length, does that change the meaning of mindist and maxdist?

Or is `dir` required to be normalized?

Or is `dir` automatically normalized internally to the trace call?

Reply:
It's is true at least for implementation in Blender, and it led to weird behavior, that's how I discovered it.

Reply:
I agree that the spec does not spell it out, and should say more fully what the behavior should be.

Personally, I think that it's really confusing for the units of min/max to depend on whether or not the direction is normalized. My preference is to fix it in Blender rather than enshrine that odd behavior in the OSL spec itself. I think it's a much simpler explanation to say that min/max are in the canonical units of the scene. (and/or, that trace expects the dir argument to be unit length)

Reply:
This behavior is going to end up being entirely under the control of the renderer integration and is largely out of the hands of this code-base, so this is more of a "do we want to suggest a uniform policy across all renderers" or "do we want to leave it up to the renderer."  The code interface that the renderer needs to implement uses Vec3 with no suggestion that they are normalized, and nothing in the current code JIT or calls will do any enforcement of normalization.

So with that, using the simplest math of P + mindist/maxdist * dir is likely what the first implementation anybody makes will do.  Irregardless of whether dir starts out normalized or not, if the renderer has to transform dir to a different space, then the renderer needs to be mindful of the scale factor of that transform and apply it to the mindist or maxdist anyway, so we're not increasing any burden there. 

It seems if we are to pick a policy to codify in the documentation, this is the most straightforward and least burdensome to implement from a renderer perspective...however the point of a domain specific language is generally to make it the least burdensome on the person developing code in that language.  How surprising do shader authors find this kind of behavior? 

FWIW, RenderMan treats the argument this way.

Reply:
I think Larry an my comments crossed paths -- Larry, there are different spaces, and the same OSL shader may run in different spaces. By "cannonical units of the scene"  -- do you mean "world" space? Or are you saying "dir" must be normalized in "common" space... i.e. the shader must call normalize on dir right before trace()...or do you mean scene units, which would be "world" space for our renderer at least.  If someone computes the ray direction in a convenient space for that, say the space defined by an orthonormal basis for the shaded point, and then transforms that to "current" it'll be non-normalized reasonably often in the trace() call, same for if it is normalized in "world" space depending on the renderer's "current" space.  

I suppose then "most convenient" for setting min/maxdist will depend on if they wanted those in "world" units or if as aribitrarily set shader parameters, or of there is some geometric bounding box or something that the shader is trying to compute an intersection with and set them up that way.  "camera" space is often convenient for setting up maxdist on depth-probe type rays.

Reply:
I think it would make sense to describe how its currently implemented, even if just noting that that is how it is done in some implementations- because a lot of implementations don't document this and just point to these specs.  It also as its a standard function I think it should have somewhat portable behavior where it isn't too difficult to implement.  Furthermore, if the behavior isn't described to a usable or implementable state what is the point in having a spec at all?

The issue comes in that I think intuitively as I discovered this writing shaders, I assumed that it was referring to world space, furthermore I think world space is more useful than some scalar, and I'm not entirely sure what scalar units would be useful for.  But, to describe that would be breaking, and not match existing implementations.  I wonder if what would be most clear is just stating that the direction vector should be normalized before passing it- or at least noting that there can be unexpected behavior if the direction vector isn't normalized.

Reply:
> If a renderer wants to allow the length of the vector to be semantically meaningful 

what I really mean here is also "If a renderer wants to optimize by not checking nor taking into account that the vector may not be normalized..."